### DISCUSSION TITLE
9p/virtfs share not writable

### DISCUSSION THREAD

**Discussion by rpolyano** [utmapp/UTM#4458] [UTM Help (macOS)] [answered]
**Describe the issue**  
9p share is not writable.

**Configuration**  
* UTM Version: 4.0.6 (65)
* macOS Version: 12.6
* Mac Chip (Intel, M1, ...): M1

**Setup**
Guest: Ubuntu Server 22.04
1) Added a VirtFS share of `/Users/roman/.../foo` in UTM
2) In guest, mounted the share with `sudo mount -t 9p -o trans=virtio share /mnt/share -o version=9p2000.L` (also tried without -version parameter)
3) `ls -l` in /mnt/share
```
ls: test.c: Network dropped connection on reset
ls: makefile: Network dropped connection on reset
total 8
-rw-rw-r-- 1 501 dialout 195 Sep 26 20:59 makefile
-rw-rw-r-- 1 501 dialout 233 Sep 26 21:00 test.c
```

I know the network dropped bug is already been reported as #4394 but my issue is that trying to modify either of these files, or create new files fails with EACCESS - `$ touch test.h --> touch: cannot touch 'test.h': Permission denied`

It appears I can make changes as root -`$ sudo touch test.h` works fine. How can I allow normal user to access the files?

**Comment by osy** [comment, marked as answer]
Try running `sudo chown -R yourusername:yourgroup path/to/share/mount`

**Reply by rpolyano** [reply]
Thank you!

**Reply by rpolyano** [reply]
Upon further use seems like only the first write works. After that they just hang. I think I'll take the advice I've seen online and give up on 9pfs, it's just not stable enough to work.

**Reply by saoj** [reply]
@rpolyano What is an alternative to 9pfs?

**Comment by sbates130272** [comment]
@osy - does your chown solution not create a new problem in that the fileownership on the host filesystem are now set to the uid and gid of the guest account in the VM?

I see exactly the same issue as @rpolyano. We can solve this by using bindfs to re-mount the 9p mount with the desired uid and gid in the guest. Something like

```
sudo bindfs --map=501/<your guest uid>:@dialout/@<your guest gid> <9p mount point> <desired bindfs mount point>
```

However it seems to me there *must* be a way this can be solved via either the call in UTM into qemo -fsdev *or* in the mount options for the mount command inside the guest. If anyone has any pointers for this I would really appreciate it.

Oh and @osy I think this is an issue rather than a discussion until we root cause a solution....

**Reply by osy** [reply]
The guest permissions are mapped to host in the file attribute, it does not change the ownership on the host file system. However different guests will see the same permissions.

**Reply by sbates130272** [reply]
I see that @osy but I also see that new files created by the guest have odd permissions set back on the host. For example if I create a file on the guest called test:

On guest
```
batesste@bunbeg:~/Projects2$ ls -la test
ls: test: Network dropped connection on reset
-rw-r--r-- 1 batesste batesste 0 Oct  6 19:23 test
```

On host
```
batesste@armagh Projects % ls -la@ test
-rw-------@ 1 batesste  staff  0  6 Oct 19:23 test
	com.apple.quarantine	25 
	user.virtfs.gid	4 
	user.virtfs.mode	2 
	user.virtfs.uid	4 
```
It is annoying that no group or global permissions are set. Any ideas if there is a better set of options in either the QEMU command or the mount command in the host?

**Comment by light-idea** [comment]
I also ran issues with permissions and solved it by installing bindfs (as above).  To get the shared folder to mount every boot, append the following to /etc/fstab:
```
share	/mnt/macos	9p	trans=virtio,version=9p2000.L,rw,_netdev,nofail	0	0
/mnt/macos    /mnt/share fuse.bindfs map=501/1000:@20/@1000,x-systemd.requires=/mnt/macos 0 0
```

**Reply by sabrehagen** [reply]
This comment should be added to the docs: https://docs.getutm.app/guest-support/linux/

**Reply by osy** [reply]
Done

**Reply by mrexodia** [reply]
I will try to make a proper guide and add it to the documentation. My goal is to mount my whole macos Home folder and running `chown -R` is **not** a good idea in that case. The magic numbers for the user/group id are also not applicable to everyone.

**Reply by mrexodia** [reply]
I sent a pull request with a better explanation (https://github.com/utmapp/docs.getutm.app/pull/45), thanks to @light-idea for findings the solution with `bindfs`!

**Reply by isolovev** [reply]
I was only able to make bindfs work by using the updated docs from the pull request. It's unfortunate that it's not yet merged into the official docs.
@osy could you maybe have a look at it?

**Reply by dmchmk** [reply]
@light-idea @mrexodia hi! How did you make 9p working? I'm getting following messages during boot:

```
[    3.518874] 9pnet: Installing 9P2000 support
[    3.521241] 9p: Installing v9fs 9p2000 file system support
[    3.528182] 9pnet_virtio: no channels available for device share
[FAILED] Failed to mount /Users
```

My /etc/fstab is the following:

```
# <file system> <dir>           <type>          <options>                       <dump>  <pass>
LABEL=rootfs    /               btrfs           defaults,ssd,compress=zstd      0       0
LABEL=docker    /var/lib/docker btrfs           defaults,ssd,compress=zstd      0       0
rosetta         /opt/rosetta    virtiofs        ro,nofail                       0       0
#share           /Users         virtiofs        rw,nofail                       0       0
share /Users 9p trans=virtio,version=9p2000.L,rw,_netdev,nofail,auto 0 0
/Users    /mnt/macos fuse.bindfs map=275111150/1000:@593637566/@1000,x-systemd.requires=/Users 0 0
```

virtiofs is working, but 9p is not(

**Reply by mrexodia** [reply]
I didn't do anything special (at least as far as I remember), but my `/etc/fstab` looks like this in case that helps:

```
# UTM Shared Folder (root)
share /mnt/macos 9p trans=virtio,version=9p2000.L,rw,_netdev,nofail,auto 0 0
# bindfs mount to remap UID/GID
/mnt/macos /home/user/macos fuse.bindfs map=502/1000:@20/@1000,x-systemd.requires=/mnt/macos,_netdev,nofail,auto 
0 0
```
